### DISCUSSION TITLE
Base16 colour scheme template, for quickly setting colours

### DISCUSSION THREAD

**Discussion by IFcoltransG** [emilk/egui#3772] [Show and tell]
I found it hard to style my egui with a colour palette, because explicitly setting all the colours for widgets, windows and etc is quite low-level, and I need something more abstract and high-level for it to be convenient. Even something like [egui-themer](https://github.com/grantshandy/egui-themer) requires setting colours manually and has clutter for styling non-colour parts of egui.

So I've put together a hasty [Base16 template](https://github.com/IFcoltransG/base16-egui) so it's quicker to colourise egui if you already have 16 colours in the [Base16](https://github.com/chriskempson/base16) format. You'd use the template and a colour scheme to generate a Rust file, put it in your project, and then call a function on your `Style` to get a new one with the colours replaced.

Feel free to PR or use the code to make new things (it's MIT-0). I haven't tested whether all the colour assignments look nice so you may see room for improvement.

**Comment by j-n-f** [comment]
Here's what I did to get this working (as it wasn't straightforward):

1. install `php` and `composer`
2. `git clone https://github.com/IFcoltransG/base16-egui`
3. `git clone https://github.com/chriskempson/base16-builder-php`
4. `cd base16-builder-php`
5. `composer install`
6. `cat /path/to/your/scheme.yaml | php base16-builder.php --template .
./base16-egui/templates/default.mustache`

It seems best to go with the original `base16-builder` as many others don't seem to implement the builder specification correctly, or do silly things like try to manage the color schemes for you (and for me they were missing the one I wanted).

Nice work, @IFcoltransG 🙂

**Reply by j-n-f** [reply]
Only other comment I would make is that some of the active/hover styles are hard to read with the default text colour.

This is really apparent with `egui::SelectableLabel`. Easy enough to change though.

This is a really great start for a crate with all the themes built in. If you just have a bunch of functions returning `Style` structs you're good to go.

**Reply by IFcoltransG** [reply]
Glad you like it, and thanks for writing down the steps you used! And sorry you had to figure it out. Unfortunately I'm not planning to maintain or develop this project in the near future, but anyone planning to take it further absolutely has my blessing.

**Reply by j-n-f** [reply]
This could just be because it's a dark theme, but I had better luck with these two changes:

```rust
visuals.widgets.active.fg_stroke.color: EXTREME_BACKGROUND,
visuals.selection.stroke.color: EXTREME_BACKGROUND,
```
